probe-rs/src/probe/usb_util.rs

```diff
@@ -10,33 +11,72 @@ pub trait InterfaceExt {
 
 impl InterfaceExt for Interface {
     fn write_bulk(&self, endpoint: u8, buf: &[u8], timeout: Duration) -> io::Result<usize> {
-        let fut = async {
-            let comp = self.bulk_out(endpoint, buf.to_vec()).await;
-            comp.status.map_err(io::Error::other)?;
+        let mut endpoint = self
+            .endpoint::<Bulk, Out>(endpoint)
+            .map_err(io::Error::from)?;
 
-            let n = comp.data.actual_length();
-            Ok(n)
+        if buf.is_empty() {
+            return Ok(0);
+        }
+
+        let mut transfer_buffer = Buffer::new(buf.len());
+        transfer_buffer.extend_from_slice(buf);
+
+        endpoint.submit(transfer_buffer);
+
+        let Some(completion) = endpoint.wait_next_complete(timeout) else {
+            endpoint.cancel_all();
```

Comment:
`cancel_all` requests the transfer to complete early with a cancelled status, but it happens asynchronously and this doesn't wait for it to finish. You'll want another `wait_next_complete` here to consume the transfer. When `endpoint` goes out of scope and drops with a pending transfer, it too requests cancellation, but it is internally reference counted and the endpoint does not become available to claim again until the cancellation finishes and releases it.

This is a potential cause of the `endpoint already in use` you're hitting in CI -- if there's something handling a timeout error and then immediately calling `read_bulk`/`write_bulk` again, you'll race the transfer cancellation for whether you can claim the endpoint again.

Reply:
I should perhaps add a convenience method for submitting and waiting for a single transfer with blocking timeout like this. This API got a lot lower-level in v0.2 because the v0.1 async API was a footgun regarding cancel-safety, but those concerns mostly don't apply to blocking. I think probe-rs wants to switch to async because it will be required for WebUSB support, though.

Reply:
I've added a step to wait for the completion after a cancel, but I'm not sure if it's necessary to explicitly surface the failure of a cancellation to complete in time so the user can opt to reset something farther up the stack. FWIW, I don't see a clean way to even differentiate that based on the error types on-hand.

Reply:
The `while let Some(cancelled)` doesn't seem right. It's possible for the transfer to return with a successful status there if the transfer completed normally after the previous `wait_next_complete` timed out but before `cancel_all` reached the host controller. In that case, this would go around the loop and `wait_next_complete` would panic with no transfer pending. 

If you're OK just dropping the returned data and returning TimedOut in that rare case, you could just ignore the result of this call.

Reply:
Good call - I spent some time digging around and thinking, and my big concern was what happens if:
* transfer issued
* timeout
* cancel_all called
* transfer result arrives
* ???

I'm preeeeeetty sure the answer is "the stack throws it away and we never find out" but dude I am a _newbie_ to USB :)